teps/0060-remote-resource-resolution.md

```diff
@@ -319,19 +324,254 @@ Why should this TEP _not_ be implemented?
+
+Users wanting to refer to tasks in git (or anywhere else) are required to figure out their own
+implementation to fetch those resources, apply them, and use them in PipelineRuns / TaskRuns.
+
+Pros:
+- Pipelines has no work to do :tada:
+- Pipelines _could_ offer some supporting docs to help people do this.
+
+Cons:
+- No opportunity to standardize around remote fetching outside the sphere of tekton bundles.
+- If Pipelines did offer docs for people constructing their own solution then they'd need to be kept up to date.
+
+### Just Implement Git Support Directly in Pipelines
+
+Bake support for referencing tasks and pipelines in Git as a first class citizen alongside
+Tekton Bundles. Add a `git.Resolver` to live in tandem with [`oci.Resolver`](https://github.com/tektoncd/pipeline/blob/main/pkg/remote/oci/resolver.go).
+
+Pros:
+- Dramatically simpler requirements than an entire remote resource resolution proposal.
+- Way less specification noise, maintenance burden, etc.
+- The only other source of tekton resources that we've had actual requests for.
+
+Cons:
+- Adds code to Pipelines that is specific to git.
+- Does not provide a path for users to integrate with other version control systems.
+- Pipelines-specific solution, where we might want to also support remote resolution in other Tekton projects.
+
+### Controller makes HTTP request to resolver to get resources
+
+Resolvers supporting different source types (git/oci/etc..) are registered with
+the Pipelines controller. The controller makes web requests to resolve remote resources:
+
+1. User submits a run type with a remote reference:
+
+```yaml
+kind: TaskRun
+metadata:
+  name: foo-run
+spec:
+  taskRef:
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    remoteType: git
+    remoteParams:
+    - name: url
+      value: https://github.com/tektoncd/catalog.git
+    - name: ref
+      value: main
+    - name: path
+      value: /tasks/git-clone/0.4/git-clone.yaml
+```
+
+2. Pipelines sends a webhook request to an endpoint registered for `git` remoteType.
+
+```json
+{
+  "url": "https://github.com/tektoncd/catalog.git",
+  "ref": "main",
+  "path": "/tasks/git-clone/0.4/git-clone.yaml"
+}
+```
+
+3. Endpoint resolves `git-clone.yaml` and returns its content in the response.
+
+```json
+{
+  "resolved": "apiVersion: tekton.dev/v1beta1\nkind: Task\nmetadata:\n  name: git-clone\nspec:\n [...]"
+}
+```
+
+Pros:
+- Writing an HTTP server that responds to requests with JSON is a relatively small development burden.
+- Possibly familiar pattern for anyone who's used Triggers' Interceptors?
+- Flexible since the endpoint can do whatever it wants before returning the resolved resource.
+- Tekton Pipelines could surface metrics around latencies, errors, etc... during resolutions.
+
+Cons:
+- Unclear how secrets would be passed for private remotes.
+- Synchronous: the resource request is a direct connection from Pipelines controller to HTTP server.
+
```

Comment:
okay maybe it's not a zany idea, seems like this might be similar to the approach pipelines as code is taking with its resolver https://github.com/openshift-pipelines/pipelines-as-code#pipelines-as-code-resolver

(it also has a "repository crd" but i dont totally follow the relationship between this and the resolver 🤔 )

Reply:
There is no relationship between PAC resolver and the PAC Repository CRD
 
The PAC resolver simply get all the Objects related to a PipelineRun and resolve everything underneath (excluding bundles) it to output a simple and unique (via generateName) PipelineRun...

The repository CRD is here to let the user be able to have a definition of the list of her Repositories and which target namespace this goes to (I can expand on the reasoning for this if you like). It is used as well to store the last statuses and other metadatas.

I can move the PAC resolver out of the PAC to experimental or somewhere else if it's useful for others, it's installable as a single command if you want to play with it.

Reply:
ah okay, thanks for the explanation @chmouel, sounds like i was jumping to some incorrect conclusions about the repository CRD.